### ISSUE TITLE
Retry policy not working?

### ISSUE THREAD

**Issue by ciror14** [gophish/gophish#3157]
Hi guys,

I've had occurrences where in the events table I have errors like:

 {"error":"Max connection attempts exceeded - dial tcp 10.100.140.39:3003: connect: connection refused"}   

I know that the receiving end of the connection has been down for an handful of seconds at most, so I'm wondering what is the retry policy for the connection? specifically what is the timeout between one failed attempt and the next one?

**Comment by glennzw** [comment]
Just to confirm; you're seeing this with SMTP email dispatches?

**Comment by ciror14** [comment]
Yes correct. My configuration is such that gophish does not send emails directly to an external smtp server but to a sort of proxy. When the "proxy" is down, I get this error. It is correct that I get connection refused but it seems that all the retry attempts are done within a second.

**Comment by glennzw** [comment]
Looking through [`models\maillog.go`](https://github.com/gophish/gophish/blob/master/models/maillog.go) we can see the following clues:

The MaxSendAttempts variable is set to 8, meaning each email will be retried up to 8 times.

The backoff mechanism increases the delay exponentially after each failed attempt. The backoff duration is calculated using `math.Pow(2, float64(m.SendAttempt))`, which means the delay will be 2^1 minutes after the first failure, 2^2 minutes after the second, and so on, up to a maximum of 256 minutes (about 4.2 hours).

When an email fails to send, the Backoff function is called. This function increments the SendAttempt counter, calculates the backoff duration, updates the SendDate to retry after the calculated delay, and then saves the updated MailLog object.

If the number of send attempts exceeds MaxSendAttempts, the email is marked as permanently failed using the ErrMaxSendAttempts error.


Hope that helps.

**Comment by ciror14** [comment]
Hi Glenn,

thanks a lot for this. I agree with your analysis, but what we experienced is that while the mechanism works in case of errors returned by the smtp server (e.g. wrong email address), if it is the connection itself to the smtp server that can not be established (e.g. connection refused) all the 8 attempts are executed with no delay resulting in an immediate "max connections attempted".

**Comment by glennzw** [comment]
Ahhhh good point, it's clearly failing at the network rather than protocol level. I'll have to give this some more thought, but suspect we'll need to tweak mailer/mailer.go:

https://github.com/gophish/gophish/blob/8e79294413932fa302212d8e785b281fb0f8896d/mailer/mailer.go#L208-L223

**Comment by Hosted-EU** [comment]
There’s A New Phishing Tool Out There.
It Can Generate URLs For Phishing Pages.
And Able To Steal Cookies from Phishing Pages To Bypass 2FA 🔥

Watch Video On How It Works : https://youtube.com/watch?v=BJv3jFuJJOw
